modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/sql/IgniteSqlParser.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.sql.engine.sql;
+
+import static org.apache.ignite.internal.util.ExceptionUtils.withCauseAndCode;
+import static org.apache.ignite.lang.ErrorGroup.extractCauseMessage;
+import static org.apache.ignite.lang.ErrorGroups.Sql.QUERY_INVALID_ERR;
+
+import java.io.Reader;
+import org.apache.calcite.config.Lex;
+import org.apache.calcite.sql.SqlNode;
+import org.apache.calcite.sql.SqlNodeList;
+import org.apache.calcite.sql.parser.SqlAbstractParserImpl;
+import org.apache.calcite.sql.parser.SqlParseException;
+import org.apache.calcite.sql.parser.SqlParser;
+import org.apache.calcite.sql.parser.SqlParserImplFactory;
+import org.apache.calcite.util.SourceStringReader;
+import org.apache.ignite.internal.generated.query.calcite.sql.IgniteSqlParserImpl;
+import org.apache.ignite.sql.SqlException;
+
+/**
+ * Implementation of SQL parser.
+ */
+public final class IgniteSqlParser extends IgniteSqlParserImpl {
```

Comment:
do we really want to expose extended parser? I would prefer to hide it inside of IgniteSqlParser as a private static class. The same with FACTORY. In that case we will have the only possible way to use new parser - through static methods `parse` as it meant to be used.

Secondly, what do you think about separating two different cases (single statement vs script) in parser as well? This let us avoid annoying validation of size of the returned nodeLIst every time. But we need to change API a bit: instead of IgniteSqlScriptNode extension we need to introduce ParsedResult encapsulating root of AST and count of dynamic parameters. As for now, code like below looks very confusing:

```                    
                    if (nodes.size() > 1) {
                        throw new SqlException(QUERY_INVALID_ERR, "Multiple statements aren't allowed.");
                    }

                    validateParsedStatement(context, nodes, nodes.get(0), params);

                    return nodes.get(0);
```

Reply:
Just to clarify - You are talking about removing `extends IgniteSqlParserImpl`  from IgniteSqlParser? Yes, I think it would be better this way.

Regarding the single/multiple statements - I agree that it would be better if single statement/multiple statement cases were using different APIs, and I can fix that.

Reply:
> Just to clarify - You are talking about removing extends IgniteSqlParserImpl from IgniteSqlParser?

yes, I'm

Reply:
Done.